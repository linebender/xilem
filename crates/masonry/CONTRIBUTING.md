# How to contribute

Issues and PRs are welcome. See [help-wanted issues](https://github.com/PoignardAzur/masonry-rs/issues?q=is%3Aissue+is%3Aopen+label%3A%22help+wanted%22) if you don't know where to begin.

## Issues

Issues don't have a specific format.

If you submit one, please be polite, remember that most Masonry contributors are unpaid volunteers, and include detailed steps to reproduce any problem.


## Pull requests

All submissions, including submissions by project members, require review. We use GitHub Pull Requests for this purpose. Consult GitHub Help for more information on using pull requests.

Before making a PR, please follow these steps:

- Run `cargo test --all-targets` on your code.
- Run `cargo test --doc`.
- Run `cargo clippy --all-targets`.

Masonry doesn't keep a changelog for now, but might in the future. Make sure your commit messages and pull request titles explain what you have changed.
How to maintain


## Preparing for a new release

Before a new release, make sure to update every dependency to their last version.

You can use `cargo upgrade` from the `cargo-edit` tool to help with the process.
