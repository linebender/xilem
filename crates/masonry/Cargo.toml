[package]
authors = ["Olivier FAURE <couteaubleu@gmail.com>"]
categories = ["gui"]
description = "Data-oriented Rust UI design toolkit."
edition = "2021"
keywords = ["gui", "ui", "toolkit"]
license = "Apache-2.0"
name = "masonry"
readme = "README.md"
repository = "https://github.com/PoignardAzur/masonry-rs"
rust-version = "1.65"
version = "0.1.3"

[profile.dev.package."*"]
opt-level = 2

# NOTE: Make sure to keep wgpu version in sync with the version badge in README.md
[dependencies]
fnv = "1.0.7"
instant = {version = "0.1.6", features = ["wasm-bindgen"]}
smallvec = "1.6.1"
tracing = "0.1.29"
tracing-subscriber = {version = "0.2.15", features = ["fmt", "ansi"], default-features = false}
image = "0.24.0"
once_cell = "1.9.0"
serde = {version = "1.0.133", features = ["derive"]}
serde_json = "1.0.74"
vello = { git = "https://github.com/linebender/vello/", rev = "b520a35addfa6bbb37d93491d2b8236528faf3b5" }
kurbo = "0.11.0"
futures-intrusive = "0.5.0"
pollster = "0.3.0"
parley = { git = "https://github.com/linebender/parley", rev = "4f05e183be9b388c6748d3c531c9ac332672fb86" }
wgpu = { version = "0.19.3" }
swash = "0.1.15"
winit = "0.29.15"

[target.'cfg(target_arch="wasm32")'.dependencies]
console_error_panic_hook = {version = "0.1.6"}
tracing-wasm = {version = "0.2.0"}

[dev-dependencies]
float-cmp = {version = "0.8.0", features = ["std"], default-features = false}
insta = {version = "1.8.0"}
assert_matches = "1.5.0"
pulldown-cmark = {version = "0.8", default-features = false}
tempfile = "3.10.1"

[target.'cfg(not(target_arch="wasm32"))'.dev-dependencies]
open = "1.6"

[[example]]
name = "simple_image"
#required-features = ["image", "png"]
