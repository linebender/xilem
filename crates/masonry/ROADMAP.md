# TODO - MVP

- [X] Clean up imports
- [X] Add impl Deref for WidgetRef
- [X] Add downcast to WidgetRef and WidgetView
- [X] Add WidgetAdded test
- [X] Rename ContextState to GlobalPassCtx
- [X] Sort WidgetState fields
- [X] Remove ExtendDrain

- [X] Rewrite tests
 - [X] Create coverage profile
 - [X] Look at coverage of every file, write a list of target coverage
 - [X] Add timers to harness

- [X] Add action queue
- [X] Re-add ExtEvent + AppDelegate

- [X] Use WidgetView everywhere
 - [X] Button
 - [X] Checkbox
 - [X] Flex
 - [X] Image
 - [X] Label
 - [X] SizedBox
 - [X] Spinner
 - [X] Portal

- [X] Implement TextBox
 - [X] Implement Scroll
 - [X] Add text/IME to harness
 - [X] Port TextBox
 - [X] Write unit tests

- [X] Refactor ScrollComponent / paint_rect / paint_insets
 - [X] Set paint_rect automatically
 - [X] Check that paint_rect has children paint_rect
 - [X] Disambiguate concept of hidden/stashed widgets

- [X] Refactor different passes
 - [X] Write a WidgetCtx trait
 - [X] Add a `WidgetPod::as_mut(ctx: &impl WidgetCtx)` method

- [X] Fix WidgetView type
 - [X] Rename WidgetView to WidgetMut
 - [X] Split WidgetRef into separate file
 - [X] Create StoreInWidgetMut from (Widget, WidgetCtx)
 - [X] Create `declare_widget!()` macro to implement boilerplate.

- [X] Write examples
 - [X] To-do List
 - [X] Fix blocking_function example
 - [X] Copy examples from druid

- [X] Fix doc tests

- [X] Split off from Druid repo
- [X] Rename to Masonry
- [X] Write remaining tasks in Github
- [X] Rewrite License bits
- [X] Write ARCHITECTURE.md
- [X] Write CONTRIBUTING.md
- [X] Write README.md and doc root
- [X] Publish

# TODO - Next steps

- [X] Remove Env type and Data trait (#8)
- [ ] Re-add Dialog feature (#25)
- [ ] Switch to using Vello and Glazier (#24)
- [ ] Refactor TextLayout (#23)

- [ ] Rework Widget trait (#26)

- [ ] Write Widget Inspector

- [ ] Port Panoramix to Masonry
- [ ] Port Xilem to Masonry

# TODO - Long term

- [ ] Brainstorm names for AppDelegate, Harness methods, Action, ExtEvent

- [ ] Move debug logger to use tracing infrastructure

- [ ] Refactor Mouse Events and hot status changes

- [ ] Remove `impl Widget for Box<dyn Widget>`

- [ ] Handle request-layout-in-layout corner case

- [ ] Add logo

- [ ] Create project board

- [ ] Refactor WidgetState
 - [ ] Add WidgetTmpState
 - [ ] Extract WidgetState fields to WidgetTmpState + global state

- [ ] Rework place_child
- [ ] Fix invalidation when computing layout

- [ ] Write analysis of different layout footguns
 - Flex interacting with viewport
 - Does Tennent's Correspondence principle apply to Wrappers in a Flex container?
 - Difference in Harness(Button) vs Harness(Flex(Button))
 - Flex alignment vs text alignment

- [ ] Refactor AppDelegate
 - Ideally, should be a single function (?)
 - Probably unify on_event, on_command, on_action

- [ ] Add WidgetBuilder trait
- [ ] Switch module file conventions (foobar/mod.rs -> _foobar.rs)
- [ ] Add dev shortcuts:
 - [ ] To print Widget tree
 - [ ] To print current focused/active Widget
 - [ ] To display invalidation rect

- [ ] Trait-ify Glazier
 - [ ] ??? Start at the bottom?

- [ ] Glazier
 - [ ] Rename WinHandler methods
  - size -> resize
  - scale -> rescale
  - command -> select_menu
  - save_as -> save_dialog_res
  - open_file -> open_dialog_res
  - timer  -> timer_complete

- [ ] Add replay mode
 - [ ] Record platform events
 - [ ] Record timers
 - [ ] Add way to rr tests

- [ ] Add ability to select rectangular region, get a list of Widget in return or something.

- [ ] Add "click all visible buttons" method to help test examples.

- [ ] Switch to ECS
- [ ] Remove WidgetId::reserved, move WidgetId::new
- [ ] Drag 'n Drop



- [ ] Add fuzzing
 - [ ] Check Masonry invariants
 - [ ] Create fuzzing suite
 - [ ] Test it with intentional bugs

- [ ] Add profiling
 - [ ] Add artinasal timings manager
 - [ ] Start in separate crate with artificial benchmarks
 - [ ] Add "median/worst FPS" log at the end of each run.

- [ ] Write benchmarks
 - [ ] Try to find ways to bust the framework's framerate
 - [ ] Huge list
 - [ ] Huge list with editable/focusable boxes
 - [ ] ???

- [ ] Refactor Command infrastructure
 - [ ] Broadcast (used multiple times)
 - [ ] Command (used once)
 - [ ] ???
 - [ ] Add on_command() trait method
 - [ ] Unify timers and promises

- [ ] Reintroduce Env type (?)
 - [ ] Remove Data trait
 - [ ] Make it so that required keys are known at launch time for better error messages.
 - [ ] Re-add Localization

- [ ] Get better coverage on Github Actions
 - Take inspiration from druid's Actions

- [ ] Improve how Harness mocks IME
 - [ ] Handle in-place editing (eg Kanji input, I think?).
 - [ ] Handle movement (eg ctrl+left).

- [ ] Merge WidgetMut and Ctx lifetimes
 - Note: probably requires better language support for implicit lifetime bounds

## Tests to write

- [X] event
- [X] promise
- [X] utils
- [X] widget_pod
 -> [X] 'active' status
 -> [X] debug_validate
 -> [X] check_initialized
 -> [X] check set_origin
 -> [X] MouseLeave
 -> [X] layout changes hot status
 -> [ ] paint_insets
 -> [ ] pan_to_child
 -> [ ] ParentWindowOrigin
 -> [ ] debug_widget_text
 -> [x] check for WidgetAdded
-> [X] widget utils
-> [X] Widgets
 -> [X] Button
 -> [X] Checkbox
 -> [X] Portal
 -> [X] Flex
 -> [X] Image
 -> [X] Label
 -> [X] SizedBox
 -> [X] Spinner
-> [ ] FillStrat
-> [ ] text
 -> [ ] TextBox


# Design stuff

What referential are various values using?

transform_scroll translates mouse events

General note: add hover detection *everywhere*. It's what makes the UI feel alive.

## Flex stuff

Flex algo
- Does Tennent's Correspondence Principle apply to Wrappers in a Flex container?
- Visualize how much "spring" each element has

Difference between
- Window(Button)
- Window(Flex(Button))

- Flex(stuff)
- Portal(Flex(stuff))

Flex alignment vs text alignment

Portal.must_fill, Portal.constrain_horizontal, Portan.constrain_vertical

In druid Spinner needs parent constraints to decide its size.
Feels like that shouldn't be allowed.

Use "*It's a Wonderful World* boardgame"-inspired pass order.

See this thread: https://xi.zulipchat.com/#narrow/stream/147922-new-members/topic/Second.20book.20example.20doesn't.20work/

How does Flex interact with Viewport?
-> use call context for debugging (wait, what does that mean again?)

Make library of commonly desired layouts
- Center object vertically
- Horizontally
- Side gutters
- Document format

How do make easily-readable test of Flex, FillStrat layout?


## Passes

Passes have access to
- on_event
-> mut self, mut ContextState, WidgetState.<layout>, mut WidgetState.<passes>, mut WidgetState.<Status>, get_widget_view()
- on_status_change
-> mut self, ContextState, WidgetState.<layout>, mut WidgetState.<passes>, WidgetState.<Status>
- lifecycle
-> mut self, mut ContextState, WidgetState.<layout>, mut WidgetState.<passes>, mut WidgetState.<Status>, get_widget_view()
- layout
-> self, mut ContextState, mut WidgetState.<layout>, WidgetState.<passes>, WidgetState.<Status>
- paint
-> self, mut ContextState, mut WidgetState.<layout>, WidgetState.<passes>, WidgetState.<Status>
- WidgetView
-> mut self, mut ContextState, WidgetState.<layout>, mut WidgetState.<passes>, mut WidgetState.<Status>, get_widget_view()

WidgetAdded and similar events are an anti-pattern. Constructors should stand on their own. (is it actually possible though?)


## Pointer status

Default(Option(cursor))
Active(Option(cursor))
Disabled
Hidden

See https://xi.zulipchat.com/#narrow/stream/147926-druid/topic/.22hot.22.20in.20the.20presence.20of.20multiple.20pointers/near/254524486


## Invariants

Druid invariants
- layout != <prev>.layout                       => request_layout() has been called
- child.origin != <prev>.child.origin           => request_layout() has been called
- <paint-calls> != <prev>.<paint-calls>         => request_paint() has been called
- children() != <prev>.children()               => children_changed is set
- !event.handled && hovered || self.has_focus   => recurse event
- lifecycle                                     => recurse lifecycle
- !is_hidden()                                  => recurse layout
- !is_hidden() && is_visible()                  => recurse paint
- is_focusable                                  => must be const
- ids are unique


## Profiling

Record spans, with:
- pass name
- widget id
- work type
- useful/redundant/testing
- include-recursed-time-in-total-time

-> percentage of work that is useful
-> what type of work


## Debug info

- Isolate subgraph of trace during errors

Event debug info:
- layout during event
- tree of visited widgets
- respective mouse pose

Widget Debuggging
- Need to see the layout tree at any given point
- Need to see which subwidgets were not rendered and why
- Need to see the properties of each widget *at render time* both graphically and in text form

## Resource handling

See swift proposal https://github.com/apple/swift-evolution/blob/main/proposals/0271-package-manager-resources.md
