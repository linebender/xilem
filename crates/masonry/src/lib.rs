// This software is licensed under Apache License 2.0 and distributed on an
// "as-is" basis without warranties of any kind. See the LICENSE file for
// details.

//! A framework that aims to provide the foundation for Rust GUI libraries.
//!
//! Masonry gives you a platform to create windows (using Glazier as a backend) each with a tree of widgets. It also gives you tools to inspect that widget tree at runtime, write unit tests on it, and generally have an easier time debugging and maintaining your app.
//!
//! The framework is not opinionated about what your user-facing abstraction will be: you can implement immediate-mode GUI, the Elm architecture, functional reactive GUI, etc, on top of Masonry.
//!
//! This project was originally was originally a fork of Druid that emerged from discussions I had with Raph Levien and Colin Rofls about what it would look like to turn Druid into a foundational library.
//!
//! ## Example
//!
//! **(TODO: FIX THIS EXAMPLE)**
//! The todo-list example looks like this:
//!
//! ```ignore
//! use masonry::widget::{prelude::*, TextBox};
//! use masonry::widget::{Button, Flex, Label, Portal, WidgetMut};
//! use masonry::Action;
//! use masonry::{AppDelegate, AppLauncher, DelegateCtx, WindowDescription, WindowId};
//!
//! const VERTICAL_WIDGET_SPACING: f64 = 20.0;
//!
//! struct Delegate {
//!     next_task: String,
//! }
//!
//! impl AppDelegate for Delegate {
//!     fn on_action(
//!         &mut self,
//!         ctx: &mut DelegateCtx,
//!         _window_id: WindowId,
//!         _widget_id: WidgetId,
//!         action: Action,
//!     ) {
//!         match action {
//!             Action::ButtonPressed => {
//!                 let mut root: WidgetMut<Portal<Flex>> = ctx.get_root();
//!                 let mut flex = root.child_mut();
//!                 flex.add_child(Label::new(self.next_task.clone()));
//!             }
//!             Action::TextChanged(new_text) => {
//!                 self.next_task = new_text.clone();
//!             }
//!             _ => {}
//!         }
//!     }
//! }
//!
//! fn main() {
//!     // The main button with some space below, all inside a scrollable area.
//!     let root_widget = Portal::new(
//!         Flex::column()
//!             .with_child(
//!                 Flex::row()
//!                     .with_child(TextBox::new(""))
//!                     .with_child(Button::new("Add task")),
//!             )
//!             .with_spacer(VERTICAL_WIDGET_SPACING),
//!     );
//!
//!     let main_window = WindowDescription::new(root_widget)
//!         .title("To-do list")
//!         .window_size((400.0, 400.0));
//!
//!     AppLauncher::with_window(main_window)
//!         .with_delegate(Delegate {
//!             next_task: String::new(),
//!         })
//!         .log_to_console()
//!         .launch()
//!         .expect("Failed to launch application");
//! }
//! ```

#![deny(
    rustdoc::broken_intra_doc_links,
    unsafe_code,
    clippy::trivially_copy_pass_by_ref
)]
#![warn(missing_docs)]
#![warn(unused_imports)]
#![allow(clippy::new_ret_no_self)]
#![allow(clippy::needless_doctest_main)]
#![allow(clippy::should_implement_trait)]
#![allow(clippy::single_match)]
#![allow(clippy::bool_assert_comparison)]
#![cfg_attr(docsrs, feature(doc_cfg))]
#![cfg_attr(not(debug_assertions), allow(unused))]

// TODO - Add logo

#[doc(inline)]
pub use kurbo;

#[macro_use]
mod util;

mod action;
mod bloom;
mod box_constraints;
mod contexts;
mod event;
pub mod paint_scene_helpers;
pub mod promise;
pub mod render_root;
pub mod testing;
pub mod text_helpers;
pub mod theme;
pub mod widget;

// TODO
pub mod app_driver;
pub mod debug_logger;
pub mod debug_values;
pub mod event_loop_runner;

pub use action::Action;
pub use box_constraints::BoxConstraints;
pub use contexts::{EventCtx, LayoutCtx, LifeCycleCtx, PaintCtx, WidgetCtx};
pub use event::{InternalLifeCycle, LifeCycle, PointerEvent, StatusChange, TextEvent, WindowTheme};
pub use kurbo::{Affine, Insets, Point, Rect, Size, Vec2};
pub use util::{AsAny, Handled};
pub use vello::peniko::{Color, Gradient};
pub use widget::{BackgroundBrush, Widget, WidgetId, WidgetPod, WidgetState};

pub use text_helpers::ArcStr;
