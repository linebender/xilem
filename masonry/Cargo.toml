[package]
name = "masonry"
version = "0.2.0"
description = "Data-oriented Rust UI design toolkit."
keywords = ["gui", "ui", "toolkit"]
categories = ["gui", "internationalization", "accessibility"]
edition.workspace = true
license.workspace = true
repository.workspace = true
homepage.workspace = true
exclude = ["/doc/", ".gitignore"]

[package.metadata.docs.rs]
all-features = true
# rustdoc-scrape-examples tracking issue https://github.com/rust-lang/rust/issues/88791
cargo-args = ["-Zunstable-options", "-Zrustdoc-scrape-examples"]

[lints]
workspace = true

[dependencies]
vello.workspace = true
wgpu.workspace = true
kurbo.workspace = true
parley.workspace = true
winit.workspace = true
smallvec.workspace = true
tracing = { workspace = true, features = ["default"] }
fnv.workspace = true
image.workspace = true
once_cell = "1.19.0"
serde = { version = "1.0.200", features = ["derive"] }
serde_json = "1.0.116"
futures-intrusive = "0.5.0"
pollster = "0.3.0"
unicode-segmentation = "1.11.0"
# TODO: Is this still the most up-to-date crate for this?
xi-unicode = "0.3.0"
tracing-subscriber = { version = "0.3.18", features = ["env-filter", "time"] }
accesskit.workspace = true
accesskit_winit.workspace = true
time = { version = "0.3.36", features = ["macros", "formatting"] }
cursor-icon = "1.1.0"
dpi.workspace = true

[target.'cfg(target_arch = "wasm32")'.dependencies]
web-time.workspace = true

[dev-dependencies]
float-cmp = { version = "0.9.0", features = ["std"], default-features = false }
image = { workspace = true, features = ["png"] }
insta = { version = "1.38.0" }
assert_matches = "1.5.0"
tempfile = "3.10.1"

[[example]]
name = "simple_image"
#required-features = ["image", "png"]
