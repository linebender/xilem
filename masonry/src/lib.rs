// Copyright 2018 the Xilem Authors and the Druid Authors
// SPDX-License-Identifier: Apache-2.0

//! A framework that aims to provide the foundation for Rust GUI libraries.
//!
//! Masonry gives you a platform to create windows (using [winit] as a backend) each with a tree of widgets. It also gives you tools to inspect that widget tree at runtime, write unit tests on it, and generally have an easier time debugging and maintaining your app.
//!
//! The framework is not opinionated about what your user-facing abstraction will be: you can implement immediate-mode GUI, the Elm architecture, functional reactive GUI, etc, on top of Masonry.
//!
//! Masonry was originally a fork of [Druid] that emerged from discussions within the Linebender community about what it would look like to turn Druid into a foundational library.
//!
//! ## Example
//!
//! The to-do-list example looks like this:
//!
//! ```
//! use masonry::app_driver::{AppDriver, DriverCtx};
//! use masonry::dpi::LogicalSize;
//! use masonry::widget::{Button, Flex, Label, Portal, RootWidget, Textbox, WidgetMut};
//! use masonry::{Action, WidgetId};
//! use winit::window::Window;
//!
//! const VERTICAL_WIDGET_SPACING: f64 = 20.0;
//!
//! struct Driver {
//!     next_task: String,
//! }
//!
//! impl AppDriver for Driver {
//!     fn on_action(&mut self, ctx: &mut DriverCtx<'_>, _widget_id: WidgetId, action: Action) {
//!         match action {
//!             Action::ButtonPressed(_) => {
//!                 let mut root: WidgetMut<RootWidget<Portal<Flex>>> = ctx.get_root();
//!                 let mut root = root.get_element();
//!                 let mut flex = root.child_mut();
//!                 flex.add_child(Label::new(self.next_task.clone()));
//!             }
//!             Action::TextChanged(new_text) => {
//!                 self.next_task = new_text.clone();
//!             }
//!             _ => {}
//!         }
//!     }
//! }
//!
//! fn main() {
//!     let main_widget = Portal::new(
//!         Flex::column()
//!             .with_child(
//!                 Flex::row()
//!                     .with_flex_child(Textbox::new(""), 1.0)
//!                     .with_child(Button::new("Add task")),
//!             )
//!             .with_spacer(VERTICAL_WIDGET_SPACING),
//!     );
//!
//!     let window_size = LogicalSize::new(400.0, 400.0);
//!     let window_attributes = Window::default_attributes()
//!         .with_title("To-do list")
//!         .with_resizable(true)
//!         .with_min_inner_size(window_size);
//!
//!     # return;
//!     masonry::event_loop_runner::run(
//!         masonry::event_loop_runner::EventLoop::with_user_event(),
//!         window_attributes,
//!         RootWidget::new(main_widget),
//!         Driver {
//!             next_task: String::new(),
//!         },
//!     )
//!     .unwrap();
//! }
//! ```
//!
//! [winit]: https://crates.io/crates/winit
//! [Druid]: https://crates.io/crates/druid

#![deny(unsafe_code, clippy::trivially_copy_pass_by_ref)]
// #![deny(rustdoc::broken_intra_doc_links)]
// #![warn(missing_docs)]
#![warn(unused_imports)]
#![allow(clippy::should_implement_trait)]
#![allow(clippy::single_match)]
#![cfg_attr(docsrs, feature(doc_cfg))]
#![cfg_attr(not(debug_assertions), allow(unused))]

// TODO - Add logo

pub use cursor_icon::{CursorIcon, ParseError as CursorIconParseError};
pub use dpi;
pub use kurbo;
pub use parley;
pub use vello;

#[macro_use]
mod util;

mod action;
mod bloom;
mod box_constraints;
mod contexts;
mod event;
pub mod paint_scene_helpers;
pub mod promise;
pub mod render_root;
pub mod testing;
// mod text;
pub mod text_helpers;
pub mod theme;
pub mod widget;

// TODO
pub mod app_driver;
pub mod debug_logger;
pub mod debug_values;
pub mod event_loop_runner;
pub mod text2;
mod tracing_backend;

pub use action::Action;
pub use box_constraints::BoxConstraints;
pub use contexts::{AccessCtx, EventCtx, LayoutCtx, LifeCycleCtx, PaintCtx, WidgetCtx};
pub use event::{
    AccessEvent, InternalLifeCycle, LifeCycle, PointerButton, PointerEvent, PointerState,
    StatusChange, TextEvent, WindowEvent, WindowTheme,
};
pub use kurbo::{Affine, Insets, Point, Rect, Size, Vec2};
pub use parley::layout::Alignment as TextAlignment;
pub use util::{AsAny, Handled};
pub use vello::peniko::{Color, Gradient};
pub use widget::{BackgroundBrush, Widget, WidgetId, WidgetPod, WidgetState};

pub use text_helpers::ArcStr;
