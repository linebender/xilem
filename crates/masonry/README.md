## Masonry

Masonry is a framework that aims to provide the foundation for Rust GUI libraries.

Masonry gives you a platform to create windows (using [Glazier](https://github.com/linebender/glazier) as a backend) each with a tree of widgets. It also gives you tools to inspect that widget tree at runtime, write unit tests on it, and generally have an easier time debugging and maintaining your app.

The framework is not opinionated about what your user-facing abstraction will be: you can implement immediate-mode GUI, the Elm architecture, functional reactive GUI, etc, on top of Masonry.

This project was originally a fork of [Druid](https://github.com/linebender/druid) that emerged from discussions I had with Raph Levien and Colin Rofls about what it would look like to turn Druid into a foundational library.

## Installing

```sh
cargo add masonry
```

### Linux

On Linux, Masonry requires gtk+3; see [GTK installation page](https://www.gtk.org/docs/installations/linux/).
(On ubuntu-based distro, running `sudo apt-get install libgtk-3-dev` from the terminal will do the job.)

### OpenBSD

On OpenBSD, Masonry requires gtk+3; install from packages:
```sh
pkg_add gtk+3
```

## Example

The todo-list example looks like this:

```rust
use masonry::widget::{prelude::*, TextBox};
use masonry::widget::{Button, Flex, Label, Portal, WidgetMut};
use masonry::Action;
use masonry::{AppDelegate, AppLauncher, DelegateCtx, WindowDescription, WindowId};

const VERTICAL_WIDGET_SPACING: f64 = 20.0;

struct Delegate {
    next_task: String,
}

impl AppDelegate for Delegate {
    fn on_action(
        &mut self,
        ctx: &mut DelegateCtx,
        _window_id: WindowId,
        _widget_id: WidgetId,
        action: Action,
    ) {
        match action {
            Action::ButtonPressed => {
                let mut root: WidgetMut<Portal<Flex>> = ctx.get_root();
                let mut flex = root.child_mut();
                flex.add_child(Label::new(self.next_task.clone()));
            }
            Action::TextChanged(new_text) => {
                self.next_task = new_text.clone();
            }
            _ => {}
        }
    }
}

fn main() {
    // The main button with some space below, all inside a scrollable area.
    let root_widget = Portal::new(
        Flex::column()
            .with_child(
                Flex::row()
                    .with_child(TextBox::new(""))
                    .with_child(Button::new("Add task")),
            )
            .with_spacer(VERTICAL_WIDGET_SPACING),
    );

    let main_window = WindowDescription::new(root_widget)
        .title("To-do list")
        .window_size((400.0, 400.0));

    AppLauncher::with_window(main_window)
        .with_delegate(Delegate {
            next_task: String::new(),
        })
        .log_to_console()
        .launch()
        .expect("Failed to launch application");
}
```

As you can see, compared to crates like Druid or Iced, Masonry takes a fairly low-level approach to GUI: there is no complex reconciliation logic or dataflow going on behind the scenes; if you want to add a widget to the flex container, you call `flex.add_child(your_widget)`.

This simplicity makes Masonry somewhat painful if you want to use it to actually build GUI applications. The hope is that, by being low-level and straightforward, developers can easily build GUI frameworks on top of it.

(Well, in theory. The first stress-test will be porting [Panoramix](https://github.com/PoignardAzur/panoramix), a React-style GUI in Rust, to Masonry.)


## Unit tests

Masonry is designed to make unit tests easy to write, as if the test function were a mouse-and-keyboard user. Tests look like this:

```rust
#[test]
fn some_test_with_a_button() {
    let [button_id] = widget_ids();
    let widget = Button::new("Hello").with_id(button_id);

    let mut harness = TestHarness::create(widget);

    // Make a snapshot test of the visual contents of the window
    assert_render_snapshot!(harness, "hello");

    harness.edit_root_widget(|mut button, _| {
        let mut button = button.downcast::<Button>().unwrap();
        button.set_text("World");
    });

    // Make new snapshot test now that the window has changed
    assert_render_snapshot!(harness, "world");

    // References to widget automatically implement Debug, and
    // will print their part of the widget hierarchy.
    println!("Window contents: {:?}", harness.root_widget());

    // You can also use insta to snapshot-test the widget hierarchy
    assert_debug_snapshot!(harness.root_widget());

    // Clicking on a button will produce a "ButtonPressed" action.
    harness.mouse_click_on(button_id);
    assert_eq!(
        harness.pop_action(),
        Some((Action::ButtonPressed, button_id))
    );
}
```

## Contributing

Issues and PRs are welcome. See [`help-wanted` issues](https://github.com/PoignardAzur/masonry-rs/issues?q=is%3Aissue+is%3Aopen+label%3A%22help+wanted%22) if you don't know where to begin.

## Roadmap

The immediate next steps are:

- [X] Remove Env type and Data trait (#8)
- [ ] Re-add Dialog feature (#25)
- [ ] Switch to using Vello and Glazier (#24)
- [ ] Refactor TextLayout (#23)

- [ ] Rework Widget trait (#26)

- [ ] Port [Panoramix](https://github.com/PoignardAzur/panoramix) to Masonry
- [ ] Port [Xilem](https://github.com/linebender/xilem) to Masonry

See [ROADMAP.md](./ROADMAP.md) and [the issues page](https://github.com/PoignardAzur/masonry-rs/issues) for more.
