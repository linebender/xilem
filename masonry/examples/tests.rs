// Copyright 2025 the Xilem Authors
// SPDX-License-Identifier: Apache-2.0

//! This module imports widget code from other examples.
//!
//! When running unit tests, it will compare the rendered snapshots of the widgets
//! with the expected snapshots.

fn main() {
    println!("This example is only used to compile other examples.");
}

#[cfg(test)]
#[allow(dead_code, reason = "We don't need to run the main functions.")]
#[allow(missing_docs, reason = "Example code doesn't need docs.")]
#[path = "."]
pub mod others {
    pub mod calc_masonry;
    pub mod custom_widget;
    pub mod grid_masonry;
    pub mod simple_image;
    pub mod to_do_list;
}

// --- MARK: TESTS ---
#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;
    use masonry::assert_render_snapshot;
    use masonry::testing::TestHarness;

    use super::others::*;

    #[test]
    fn calc_masonry_screenshot() {
        let mut harness = TestHarness::create(calc_masonry::build_calc());
        assert_debug_snapshot!(harness.root_widget());
        assert_render_snapshot!(harness, "calc_masonry_initial");

        // TODO - Test clicking buttons
    }

    #[test]
    fn custom_widget_screenshot() {
        let my_string = "Masonry + Vello".to_string();

        let mut harness = TestHarness::create(custom_widget::CustomWidget(my_string));
        assert_debug_snapshot!(harness.root_widget());
        assert_render_snapshot!(harness, "custom_widget_initial");
    }

    #[test]
    fn grid_masonry_screenshot() {
        let mut harness = TestHarness::create(grid_masonry::make_grid(1.0));
        assert_debug_snapshot!(harness.root_widget());
        assert_render_snapshot!(harness, "grid_masonry_initial");

        // TODO - Test clicking buttons
    }

    #[test]
    fn simple_image_screenshot() {
        let mut harness = TestHarness::create(simple_image::make_image());
        assert_debug_snapshot!(harness.root_widget());
        assert_render_snapshot!(harness, "simple_image_initial");
    }

    #[test]
    fn to_do_list_screenshot() {
        let mut harness = TestHarness::create(to_do_list::make_widget_tree());
        assert_debug_snapshot!(harness.root_widget());
        assert_render_snapshot!(harness, "to_do_list_initial");

        // TODO - Test clicking buttons
        // TODO - Test typing text
    }
}
