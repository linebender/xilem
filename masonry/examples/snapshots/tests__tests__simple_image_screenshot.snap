---
source: masonry/examples/tests.rs
assertion_line: 57
expression: harness.root_widget()
---
Image
